address the issue #84: potential mistake in two cost functions

chengsijin0817 opened the pull request.

hartig commented on src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/costmodel/CFRNumberOfTermsShippedInResponses.java:
```diff
@@ -61,26 +58,19 @@ else if ( lop instanceof LogicalOpBGPAdd ) {
+			}
 			else {
-				throw createIllegalArgumentException(req);
+				throw createIllegalArgumentException(fm);
```
Before throwing this exception, `futureIntResSize` should be canceled. See the case of `LogicalOpTPAdd` a bit further up in this source file.

hartig commented on src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/costmodel/CFRNumberOfVarsShippedInResponses.java:
```diff
@@ -59,24 +56,19 @@ else if ( lop instanceof LogicalOpBGPAdd ) {
 		}
 		else if ( lop instanceof LogicalOpRequest ) {
-			final DataRetrievalRequest req = ((LogicalOpRequest<?, ?>) lop).getRequest();
+			final CompletableFuture<Integer> futureIntResSize = initiateCardinalityEstimation(plan);
```
In this case, `futureIntResSize` should be created not here but only within the if-branch for SPARQL endpoints. The reason for this is that `futureIntResSize` is not needed for TPF/brTPF servers. We shouldn't start a cardinality estimation process if we don't actually need the cardinality estimate ;-)

hartig commented on src/main/java/se/liu/ida/hefquin/engine/query/impl/QueryPatternUtils.java:
```diff
@@ -76,6 +76,17 @@ else if ( p instanceof BGP ) {
 	}
 
+	public static Set<Var> getVariablesInPattern( final SPARQLGraphPattern queryPattern ) {
+		Set<Var> result = new HashSet<>();
+		if ( queryPattern instanceof TriplePattern) {
+			result = getVariablesInPattern( (TriplePattern) queryPattern );
+		}
+		else if ( queryPattern instanceof BGP) {
+			result = getVariablesInPattern( (BGP) queryPattern);
+		}
+		return result;
```
```suggestion
		if ( queryPattern instanceof TriplePattern ) {
			return getVariablesInPattern( (TriplePattern) queryPattern );
		}
		else if ( queryPattern instanceof BGP ) {
			return getVariablesInPattern( (BGP) queryPattern );
		}
		else {
			throw new UnsupportedOperationException("Getting the variables from arbitrary SPARQL patterns is an open TODO (type of Jena Op in the current case: " + queryPattern.asJenaOp().getClass().getName() + ").");
		}
```

hartig replied:
No need to create a HashSet that is never used ;-)

hartig reviewed (changes requested):
Thanks for addressing the issue!

There are a few minor things that need to be changed; see my detailed comments.

Additionally, I notice that this PR is not only addressing #84, but it also addresses #85 to 50%. Since you have started it already, can you please address the remaining part of #85 as well within this PR ;-)

chengsijin0817 commented:
> Thanks for addressing the issue!
> 
> There are a few minor things that need to be changed; see my detailed comments.

Fixed. Please check it.

> 
> Additionally, I notice that this PR is not only addressing #84, but it also addresses #85 to 50%. Since you have started it already, can you please address the remaining part of #85 as well within this PR ;-)

Issue #85 is also addressed whin this PR.

hartig reviewed (approved):
Perfect. Thanks!
